pandas/tests/frame/test_alter_axes.py

```diff
@@ -28,244 +25,287 @@
+    def test_set_index_drop_inplace(self, drop, inplace, keys):
+        df = self.dummy.copy()
+
+        if isinstance(keys, list):
+            idx = MultiIndex.from_arrays([df[x] for x in keys], names=keys)
+        else:
+            idx = Index(df[keys], name=keys)
+        expected = df.drop(keys, axis=1) if drop else df
+        expected.index = idx
+
+        if inplace:
+            result = df.copy()
+            result.set_index(keys, drop=drop, inplace=True)
+        else:
+            result = df.set_index(keys, drop=drop)
+
+        tm.assert_frame_equal(result, expected)
+
+    # A has duplicate values, C does not
+    @pytest.mark.parametrize('keys', ['A', 'C', ['A', 'B']])
+    @pytest.mark.parametrize('drop', [True, False])
+    def test_set_index_append(self, drop, keys):
+        df = self.dummy.copy()
+
+        keys = keys if isinstance(keys, list) else [keys]
+        idx = MultiIndex.from_arrays([df.index] + [df[x] for x in keys],
+                                     names=[None] + keys)
+        expected = df.drop(keys, axis=1) if drop else df.copy()
+        expected.index = idx
+
+        result = df.set_index(keys, drop=drop, append=True)
+
+        tm.assert_frame_equal(result, expected)
+
+    # A has duplicate values, C does not
+    @pytest.mark.parametrize('keys', ['A', 'C', ['A', 'B']])
+    @pytest.mark.parametrize('drop', [True, False])
+    def test_set_index_append_to_multiindex(self, drop, keys):
+        # append to existing multiindex
+        df = self.dummy.set_index(['D'], drop=drop, append=True)
 
-        expected = df.loc[:, ['C', 'D', 'E']]
-        expected.index = index
+        keys = keys if isinstance(keys, list) else [keys]
+        expected = self.dummy.set_index(['D'] + keys, drop=drop, append=True)
 
-        expected_nodrop = df.copy()
-        expected_nodrop.index = index
+        result = df.set_index(keys, drop=drop, append=True)
 
-        assert_frame_equal(result, expected)
-        assert_frame_equal(result_nodrop, expected_nodrop)
-        assert result.index.names == index.names
+        tm.assert_frame_equal(result, expected)
 
-        # inplace
-        df2 = df.copy()
-        df2.set_index(['A', 'B'], inplace=True)
-        assert_frame_equal(df2, expected)
+    def test_set_index_after_mutation(self):
+        # GH1590
+        df = DataFrame({'val': [0, 1, 2], 'key': ['a', 'b', 'c']})
+        expected = DataFrame({'val': [1, 2]},
+                             Index(['b', 'c'], name='key'))
 
-        df3 = df.copy()
-        df3.set_index(['A', 'B'], drop=False, inplace=True)
-        assert_frame_equal(df3, expected_nodrop)
+        df2 = df.loc[df.index.map(lambda indx: indx >= 1)]
+        result = df2.set_index('key')
+        tm.assert_frame_equal(result, expected)
+
+    # also test index name if append=True (name is duplicate here for B)
+    @pytest.mark.parametrize('box', [Series, Index, np.array, 'MultiIndex'])
```

Comment:
dont' use a string, just directly put the lambda here. ideally we have NO if/else in test functions, sometimes its unavoidable, but making tests as simple as possible is the key